src/lang/checkers/Cashflow.ml

```diff
@@ -0,0 +1,1313 @@
+    | Money        , Money         -> Money
+    | NotMoney     , NotMoney      -> NotMoney
+    | _            , _             -> Inconsistent
+
+  (* Greatest lower bound in the money_tag lattice. *)
+  let rec glb_tags t1 t2 =
+    match t1, t2 with
+    | Inconsistent , x
+    | x            , Inconsistent  -> x
+    | NoInfo       , _
+    | _            , NoInfo        -> NoInfo
+    | Map x        , Map y         -> Map (glb_tags x y)
+    | Option x     , Option y      -> Option (glb_tags x y)
+    | Pair (x1, x2), Pair (y1, y2) -> Pair (glb_tags x1 y1, glb_tags x2 y2)
+    | Money        , Money         -> Money
+    | NotMoney     , NotMoney      -> NotMoney
+    | _            , _             -> NoInfo
+
+  (*******************************************************)
+  (*      Helper functions for local variables           *)
+  (*******************************************************)
+
+  let get_id_tag id =
+    match id with
+    | Ident (_, (tag, _)) -> tag
+      
+  let update_id_tag id new_tag =
+    match id with
+    | Ident (v, (_, rep)) -> Ident (v, (new_tag, rep))
+
+  let lookup_var_tag i env =
+    match AssocDictionary.lookup (get_id i) env with
+    | Some t -> t
+    | None -> get_id_tag i
+  
+  let update_ids_tags ids env =
+    List.map
+      (fun i ->
+         let i_tag = lookup_var_tag i env in
+         update_id_tag i i_tag) ids
+
+  let lookup_var_tag2 i env1 env2 =
+    match AssocDictionary.lookup (get_id i) env1 with
+    | Some t -> t
+    | None -> lookup_var_tag i env2
+
+  let update_var_tag2 i t env1 env2 =
+    match AssocDictionary.lookup (get_id i) env1 with
+    | Some _ -> (AssocDictionary.update (get_id i) t env1, env2)
+    | None -> (env1, AssocDictionary.update (get_id i) t env2)
+
+  (*******************************************************)
+  (*      Helper functions for builtin functions         *)
+  (*******************************************************)
+
+  (* Calculate the signature of a builtin function.
+
+     Step 1: Calculate candidate signatures based on the 
+     desired result tag and each argument tag.
+
+     - For each tag t, pick every least upper bound of that 
+     tag that makes sense for that result/argument. 
+     Call these bounds b_t.
+
+     - For each b_t, find all sets of tags satisfying that 
+     the use of those tags in the other argument/result positions 
+     is the greatest lower bound of a consistent use of tags 
+     satisfying b_t. These sets along with b_t are considered 
+     the candidate sigantures for t, called C_t.
+
+     Step 2: Consider the elements of C_t1 x C_t2 x ..., i.e., 
+     the cartesian product of the candidate signature sets for 
+     each t.
+
+     For each element, calculate the least upper bound of all 
+     the tags in the signatures of the element. Call the 
+     resulting set of candidate signatures C.
+
+     Step 3: Calculate the greatest lower bound of C. *)
+  let builtin_signature f res_tag args_tags =
```

Comment:
Is there any way to factor this sheet of tag assignments into a file that can be parsed independently? As of now, it seems like a lot this can just go into a text table.

Reply:
The table sort of belongs in Builtin.ml, but the `money_tag` datatype needs to be visible to the function, and that datatype belongs in Cashflow.ml.

I can move the function to a separate file if you want, but it would be for purely aesthetic reasons. Is it worth it?